[Feature] Add dig method to CSV::Table and CSV::Row

tomog105 opened the pull request:
## Abstract
- This PR implements `dig` method to CSV::Table and CSV::Row

## Background
- Already, Array, Hash and Struct class has `dig` method from Ruby 2.3, but CSV::Table and CSV::Row not has this method.
- I would like to use `#dig` for CSV::Table or nested arrays (or hashes) with CSV::Row without distinction.

## Usage

```rb
arr_of_arrs = CSV.parse("first,second,third\n1,2,3\n", headers: true)

# as single value
arr_or_arrs.dig(0) # row mode
=> #<CSV::Row "first":"1" "second":"2" "third":"3">
arr_of_arrs.dig("first") # col mode
=> ["1"]

# as nested values
arr_of_arrs.dig(0, "first")
=> "1"
arr_of_arrs.dig("first", 0)
=> "1"
```

kou commented on lib/csv.rb:
```diff
@@ -903,6 +918,21 @@ def to_csv(write_headers: true, **options)
+    # returning nil if any intermediate step is nil.
+    #
+    def dig(index_or_header, *index_or_headers)
+      value = send(:[], index_or_header)
```
`self[index_or_header` is better than `send(:[], index_or_header)`.

tomog105 replied:
Thanks, I fixed this part.

kou commented on test/csv/test_row.rb:
```diff
@@ -377,4 +377,16 @@ def test_can_be_compared_when_not_a_row
+
+    # if missing
+    assert_nil(@row.dig("Missing"))
+    assert_nil(@row.dig(100))
```
Can you add a test for multiple arguments case?

tomog105 replied:
I added multiple arguments case.

kou commented on lib/csv.rb:
```diff
@@ -547,6 +547,21 @@ def to_csv(**options)
+    # Extracts the nested value specified by the sequence of name objects by calling dig at each step,
+    # returning nil if any intermediate step is nil.
+    #
+    def dig(index_or_header, *index_or_headers)
```
`*index_or_headers` isn't a good name because they are arguments for `Array#dig`. `Array` doesn't have header.
`*indexes` is better.

tomog105 replied:
Thanks, I corrected arguments name.

kou commented on lib/csv.rb:
```diff
@@ -547,6 +547,21 @@ def to_csv(**options)
     alias_method :to_s, :to_csv
 
+    #
+    # Extracts the nested value specified by the sequence of name objects by calling dig at each step,
```
I don't think that `the sequence of name objects` is good description.
`the +index+ or +header+ object and the followed sequence of +index+ objects` (this English may be broken) or something will be better.

kou commented on lib/csv.rb:
```diff
@@ -903,6 +918,21 @@ def to_csv(write_headers: true, **options)
     alias_method :to_s, :to_csv
 
+    #
+    # Extracts the nested value specified by the sequence of name objects by calling dig at each step,
```
Same here.

kou commented on test/csv/test_table.rb:
```diff
@@ -494,4 +494,49 @@ def test_inspect_encoding_is_ascii_compatible
+
+    # by cell, row then col
+    assert_equal(2, @table.dig(0, 1))
+    assert_equal(6, @table.dig(1, "C"))
```
Can you add tests for three or more arguments case?

tomog105 replied:
I added three or more arguments case.

tomog105 commented:
I fixed about the part reviewed.
and I changed behavior when following value does not have `#dig` method.
(would raise same error as Array and Hash `#dig`.)

kou commented:
Thanks.
I've merged with some modifications.
